File: Array/2406.Divide_Intervals_Into_Minimum_Number_of_Groups/C++/2406.Divide_Intervals_Into_Minimum_Number_of_Groups.cpp
class Solution {
public:
    int minGroups(vector<vector<int>> & intervals) 
    {
        //=================================================//
        // Input type:                                     //
        //   - vector<vector<int>> ref. (intervals vector) //
        // Return type:                                    //            
        //   - int (result minimun groups)                 //
        //=================================================//

        /*Initialize*/
        ///// Length of intervals vector /////
        int len_intervals = intervals.size();

        ///// Record heap, priority-queue /////
        priority_queue<int, vector<int>, greater<int>> record_pq;


        /*Heap, Priority-queue based loop traversal with sorted-operations*/
        sort(intervals.begin(), intervals.end(), [&](const vector<int> & vect1, const vector<int> & vect2)
        {
            return (vect1[0] < vect2[0]);

        }); //Sorted-operations

        for (int intervals_idx = 0; (intervals_idx < len_intervals); (intervals_idx++)) //Whole
        {
            ///// Check if the current indexed-value matched conditions or not /////
            if ((! record_pq.empty()) && ((intervals[intervals_idx])[0] > record_pq.top()))
            {
                record_pq.pop(); //Keep updating, popped
            }
            else { ; }

            record_pq.emplace((intervals[intervals_idx])[1]); //Keep updating, recording

        } //Whole

        return record_pq.size();
    }
};

File: Array/2406.Divide_Intervals_Into_Minimum_Number_of_Groups/Python/2406.Divide_Intervals_Into_Minimum_Number_of_Groups.py
class Solution(object):
    def minGroups(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: int
        """
        #============================================#
        # Heap/Priority-queue based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of intervals array #####
        len_intervals = len(intervals)

        ##### Record heap/priority-queue #####
        record_pq = []


        #################################################################
        #Heap/Priority-queue based loop traversal with sorted-operations
        heapify(record_pq) #Update/Heapify

        intervals.sort() #Sorted-operations

        for intervals_idx in range(len_intervals):

            ##### Check if the current indexed-value matched conditions or not #####
            if (record_pq and (record_pq[0] < (intervals[intervals_idx])[0])):
                heappop(record_pq) #Keep updating/popped
            else:
                pass

            heappush(record_pq, (intervals[intervals_idx])[1]) #Keep updating/recording

        return len(record_pq)
